akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala

```diff
@@ -512,6 +520,20 @@ private[stream] class ActorGraphInterpreter(_initial: GraphInterpreterShell) ext
         false
     }
 
+  //this limits number of messages that can be processed synchronously during one actor receive.
+  private val shortCircuitLimit: Int = _initial.mat.settings.syncProcessingLimit
+  private var shortCircuitBuffer: FixedSizeBuffer[BoundaryEvent] = null
+  private var currentLimit: Int = shortCircuitLimit
+
+  def enqueueToShortCircuit(input: AsyncInput): Boolean = {
+    if (shortCircuitBuffer == null) shortCircuitBuffer = FixedSizeBuffer[BoundaryEvent](shortCircuitLimit)
+    if (!shortCircuitBuffer.isFull && currentLimit > 0) {
```

Comment:
we have several cases here:
1)  asyncInput targeting for another actor - we use standard actor mechanism
2) asyncInput is for current actor and the limit is not yet reached. We need to enqueue this into shortCircuitBuffer and ran after general handler finished
3) asyncInput is for current actor and it's already in shot circuit loop. Here we can enqueue asyncInput only in case if we haven't reached limit yet. 

My point is that if we are sending AsyncInput to different actors we cannot guarantee order here and shortCircuit does not work for this case. When we only dealing with one actor. We can enqueue messages until limit not reached. Until that moment we supporting message order because we are using shortCircuitBuffer. When limit is reached (once per actor.receive) we know we need to interrupt this thread to let other message come and we are sending all messages via actor mechanism - preserving message order as well.